staging/src/k8s.io/client-go/tools/cache/mutation_detector_test.go

```diff
@@ -0,0 +1,80 @@
+// +build !race
```

Comment:
does anyone remember why the race detector is turned off on this test?

Reply:
to answer my own question, because there is a data race:
```
==================
WARNING: DATA RACE
Read at 0x00c42026a540 by goroutine 89:
  reflect.maplen()
      GOROOT/src/runtime/hashmap.go:1232 +0x0
  reflect.Value.Len()
      GOROOT/src/reflect/value.go:1047 +0x1f1
  reflect.deepValueEqual()
      GOROOT/src/reflect/deepequal.go:110 +0xc0b
  reflect.deepValueEqual()
      GOROOT/src/reflect/deepequal.go:101 +0x7ea
  reflect.deepValueEqual()
      GOROOT/src/reflect/deepequal.go:101 +0x7ea
  reflect.deepValueEqual()
      GOROOT/src/reflect/deepequal.go:98 +0x114e
  reflect.DeepEqual()
      GOROOT/src/reflect/deepequal.go:196 +0x2bf
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*defaultCacheMutationDetector).CompareObjects()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector.go:116 +0x15b
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*defaultCacheMutationDetector).Run()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector.go:85 +0x50
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(CacheMutationDetector).Run-fm()
      staging/src/k8s.io/client-go/tools/cache/shared_informer.go:219 +0x50
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel.func1()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:54 +0x45
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start.func1()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:71 +0x5c

Previous write at 0x00c42026a540 by goroutine 18:
  runtime.mapassign_faststr()
      GOROOT/src/runtime/hashmap_fast.go:694 +0x0
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.TestMutationDetector()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector_test.go:73 +0x7dd
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 89 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).Start()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:69 +0x6f
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.(*Group).StartWithChannel()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:53 +0xc7
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).Run()
      staging/src/k8s.io/client-go/tools/cache/shared_informer.go:219 +0x660

Goroutine 18 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
  testing.runTests()
      GOROOT/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:978 +0x2cd
  main.main()
      bazel-out/k8-fastbuild/bin/staging/src/k8s.io/client-go/tools/cache/linux_amd64_race_stripped/go_default_test%/testmain.go:208 +0x2e9
==================
==================
WARNING: DATA RACE
Read at 0x00c42028f838 by goroutine 88:
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/apis/meta/v1.(*ObjectMeta).DeepCopyInto()
      staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/zz_generated.deepcopy.go:653 +0xd9c
  k8s.io/kubernetes/vendor/k8s.io/api/core/v1.(*Pod).DeepCopyInto()
      staging/src/k8s.io/api/core/v1/zz_generated.deepcopy.go:2990 +0x10e
  k8s.io/kubernetes/vendor/k8s.io/api/core/v1.(*Pod).DeepCopy()
      staging/src/k8s.io/api/core/v1/zz_generated.deepcopy.go:3002 +0x5c
  k8s.io/kubernetes/vendor/k8s.io/api/core/v1.(*Pod).DeepCopyObject()
      staging/src/k8s.io/api/core/v1/zz_generated.deepcopy.go:3008 +0x38
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*defaultCacheMutationDetector).AddObject()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector.go:102 +0xd2
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).HandleDeltas()
      staging/src/k8s.io/client-go/tools/cache/shared_informer.go:353 +0x1d0
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).HandleDeltas-fm()
      staging/src/k8s.io/client-go/tools/cache/shared_informer.go:202 +0x55
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*DeltaFIFO).Pop()
      staging/src/k8s.io/client-go/tools/cache/delta_fifo.go:444 +0x37c
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).processLoop()
      staging/src/k8s.io/client-go/tools/cache/controller.go:150 +0x6d
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).(k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.processLoop)-fm()
      staging/src/k8s.io/client-go/tools/cache/controller.go:124 +0x41
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil.func1()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:133 +0x61
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.JitterUntil()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:134 +0xcd
  k8s.io/kubernetes/vendor/k8s.io/apimachinery/pkg/util/wait.Until()
      staging/src/k8s.io/apimachinery/pkg/util/wait/wait.go:88 +0x5a
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*controller).Run()
      staging/src/k8s.io/client-go/tools/cache/controller.go:124 +0x3fd
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.(*sharedIndexInformer).Run()
      staging/src/k8s.io/client-go/tools/cache/shared_informer.go:227 +0x75d

Previous write at 0x00c42028f838 by goroutine 18:
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.TestMutationDetector()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector_test.go:73 +0x7f3
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 88 (running) created at:
  k8s.io/kubernetes/vendor/k8s.io/client-go/tools/cache.TestMutationDetector()
      staging/src/k8s.io/client-go/tools/cache/mutation_detector_test.go:65 +0x6f5
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d

Goroutine 18 (running) created at:
  testing.(*T).Run()
      GOROOT/src/testing/testing.go:824 +0x564
  testing.runTests.func1()
      GOROOT/src/testing/testing.go:1063 +0xa4
  testing.tRunner()
      GOROOT/src/testing/testing.go:777 +0x16d
  testing.runTests()
      GOROOT/src/testing/testing.go:1061 +0x4e1
  testing.(*M).Run()
      GOROOT/src/testing/testing.go:978 +0x2cd
  main.main()
      bazel-out/k8-fastbuild/bin/staging/src/k8s.io/client-go/tools/cache/linux_amd64_race_stripped/go_default_test%/testmain.go:208 +0x2e9
==================
```

I wonder if this is fixable?

Reply:
The cache mutation detector is constantly retrieving data from the shared informer's cache to detect for changes from external actors, without using locks. It should only ever be enabled in unit tests. I think this is working as designed.